jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/RetainableByteBuffer.java

```diff
@@ -159,6 +170,419 @@ default void clear()
+            if (newCapacity > _maxCapacity)
+            {
+                newCapacity = Math.addExact(capacity, spaceNeeded - space);
+                if (newCapacity > _maxCapacity)
+                    newCapacity = _maxCapacity;
+            }
+
+            RetainableByteBuffer ensured = _pool.acquire(newCapacity, _direct);
+            ensured.append(_buffer);
+            _buffer.release();
+            _buffer = ensured;
+        }
+    }
+
+    /**
+     * Obtain a new accumulating {@link RetainableByteBuffer} that may internally accumulate multiple other
+     * {@link RetainableByteBuffer}s with zero-copy if the {@link #append(RetainableByteBuffer)} API is used
+     * @param bufferPool The pool from which to allocate buffers
+     * @param direct true if direct buffers should be used
+     * @param maxLength The maximum length of the accumulated buffers or -1 for no limit
+     * @return The accumulating {@link RetainableByteBuffer}
+     */
+    static RetainableByteBuffer newAccumulator(ByteBufferPool bufferPool, boolean direct, int maxLength)
+    {
+        return new Accumulator(bufferPool, direct, maxLength);
+    }
+
+    class Accumulator implements RetainableByteBuffer
+    {
+        private final ByteBufferPool _pool;
+        private final boolean _direct;
+        private final long _maxLength;
+        private final List<RetainableByteBuffer> _buffers = new ArrayList<>();
+        private boolean _lastAcquiredByUs;
+
+        private Accumulator(ByteBufferPool pool, boolean direct, long maxLength)
+        {
+            _pool = pool == null ? new ByteBufferPool.NonPooling() : pool;
+            _direct = direct;
+            _maxLength = maxLength < 0 ? Long.MAX_VALUE : maxLength;
+        }
+
+        @Override
+        public ByteBuffer getByteBuffer()
+        {
+            return switch (_buffers.size())
+            {
+                case 0 -> RetainableByteBuffer.EMPTY.getByteBuffer();
+                case 1 -> _buffers.get(0).getByteBuffer();
+                default ->
+                {
+                    int length = remaining();
+                    RetainableByteBuffer combinedBuffer = _pool.acquire(length, _direct);
+                    ByteBuffer byteBuffer = combinedBuffer.getByteBuffer();
+                    BufferUtil.flipToFill(byteBuffer);
+                    for (RetainableByteBuffer buffer : _buffers)
+                    {
+                        buffer.putTo(byteBuffer);
+                        buffer.clear();
+                        buffer.release();
+                    }
+                    BufferUtil.flipToFlush(byteBuffer, 0);
+                    _buffers.clear();
+                    _buffers.add(combinedBuffer);
+                    _lastAcquiredByUs = true;
+                    yield combinedBuffer.getByteBuffer();
+                }
+            };
+        }
+
+        @Override
+        public int remaining()
```

Comment:
I'm worried about the long-to-int conversion. We should carefully weight and document if we throw or cap on int overflow.

Reply:
I'm wondering if we shouldn't cap the accumulator to 2GB?